src/HttpServer.php

```diff
@@ -60,50 +64,115 @@ public function addMiddleware($middleware)
             }
+        });
+    }
 
-            if (isset($request)) {
-                if ($that->isChunkedEncodingActive($request)) {
-                    $chunkedDecoder->on('data', function ($chunk) use (&$bodyBuffer, &$request, $that, $connection) {
-                        $bodyBuffer .= $chunk;
-                        if (strlen($chunk) == 0) {
-                            $that->sendBody($bodyBuffer, $connection, $request);
-                        }
-                    });
-                    $chunkStream->emit('data', array($data));
-                } else {
-                    $bodyBuffer .= $data;
-                    $contentLengthArray = $request->getHeader('Content-Length');
-
-                    if (!empty($contentLengthArray) && strlen($bodyBuffer) == $contentLengthArray[0]) {
-                        $that->sendBody($bodyBuffer, $connection, $request);
-                    } else if (empty($contentLengthArray) || $contentLengthArray[0] == 0) {
-                        $that->handleRequest($connection, $request);
-                    }
+    /** @internal */
+    public function handleBody(RequestInterface $request, ConnectionInterface $connection)
+    {
+        $input = $this->sendHttpBodyStream($request, $connection);
+
+        $connection->removeAllListeners('data');
+        $that = $this;
+
+        if ($this->isChunkedEncodingActive($request)) {
+            $connection->on('data', function ($data) use ($input){
+                $input->emit('data', array($data));
+            });
+            return;
+        }
+
+        if (!$request->hasHeader('Content-Length')) {
+            $connection->on('data', function ($data) use ($that, $input, $connection){
+                if ($data === '') {
+                    // Simple request without body and without 'Content-Length'
+                    $input->emit('end', array());
+                    return;
                 }
+                // Request hasn't defined 'Content-Length' and body given
+                $that->sendResponse(new Response(411), $connection);
+            });
+            return;
+        }
+
+        $contentLength = $request->getHeaderLine('Content-Length');
+
+        $int = (int) $contentLength;
+        if ((string)$int !== (string)$contentLength) {
+            // Send 400 status code if the value of 'Content-Length'
+            // is not an integer or is duplicated
+            $this->sendResponse(new Response(400), $connection);
+            return;
+        }
+
+        $transferredLength = 0;
+
+        $connection->on('data', function ($data) use ($that, $contentLength, &$transferredLength, $input) {
+            // 'Content-Length' is given
+            if (($transferredLength + strlen($data)) > $contentLength) {
+                // Only emit data until the value of 'Content-Length' is reached, the rest will be ignored
+                $data = substr($data, 0, $contentLength - $transferredLength);
```

Comment:
What if the connection keeps emitting more and more data? We should probably add some safeguards here